### PR TITLE
[FIX] account_invoice_margin, account_invoice_margin_sale: add generice method for better inheritance

### PR DESCRIPTION
**Pull request by bodedra**

### CODE REVIEW

**Comment by bodedra** [comment]
@sergio-teruel @pedrobaeza fast-track?

**Comment by sergio-teruel** [comment]
Sorry @bodedra but account_invoice_margin in v13.0 has not sale dependency...
Are you sure that you has lastest code??

https://github.com/OCA/margin-analysis/blob/13.0/account_invoice_margin/__manifest__.py#L16

**Comment by bodedra** [comment]
Yes. Correct. That's why I proposed this PR to add `sale` dependency.

`sale_line_ids` declared in sale module and we are used field at https://github.com/OCA/margin-analysis/blob/13.0/account_invoice_margin/models/account_invoice.py#L44

So it breaks my unit test with `AttributeError: 'account.move.line' object has no attribute 'sale_line_ids'`

**Comment by pedrobaeza** [comment]
@sergio-teruel you added such hidden dependency in #72

**Comment by sergio-teruel** [comment]
Yes... A mistake!!
I am thinking that the better solution is move this condition to account_invoice_margin_sale.
Mybe you can add a method to filter records and in account_invoice_margin_sale module do a filter by downpayment..

What are you think about this?

**Comment by bodedra** [comment]
We can not call super method in account_invoice_margin_sale. We have to rewrite entire method, without super. Will it work for you?

**Comment by pedrobaeza** [comment]
A possible solution is:
```python
    def _get_margin_applicable_lines(self):
        self.ensure_one()
        return self.invoice_line_ids

    @api.depends(
        "invoice_line_ids.margin",
        "invoice_line_ids.margin_signed",
        "invoice_line_ids.price_subtotal",
    )
    def _compute_margin(self):
        for invoice in self:
            margin = 0.0
            margin_signed = 0.0
            price_subtotal = 0.0
            for line in invoice._get_margin_applicable_lines():
                margin += line.margin
                margin_signed += line.margin_signed
                price_subtotal += line.price_subtotal
            invoice.margin = margin
            invoice.margin_signed = margin_signed
            invoice.margin_percent = (
                price_subtotal and margin_signed / price_subtotal * 100 or 0.0
            )
```
and in `account_invoice_margin_sale`:
```python
    def _get_margin_applicable_lines(self):
        lines = super()._get_margin_applicable_lines()
        return lines.filtered(lambda x: not x.is_downpayment)
```

File: account_invoice_margin/models/account_invoice.py [hunk 1]
```diff
@@ -40,7 +44,7 @@ def _compute_margin(self):
             margin = 0.0
             margin_signed = 0.0
             price_subtotal = 0.0
-            for line in invoice.invoice_line_ids:
+            for line in invoice._get_margin_applicable_lines():
                 if any(line.sale_line_ids.mapped("is_downpayment")):
```
Comment (pedrobaeza) [inline account_invoice_margin/models/account_invoice.py]: You should remove this and if you can do the other change as well in the same PR.

Comment (bodedra) [reply account_invoice_margin/models/account_invoice.py]: Yes. I am on it.

**Comment by pedrobaeza** [comment]
What about `account_invoice_margin_sale`?

**Comment by bodedra** [comment]
@pedrobaeza Would you like two PR or one PR with two module change?

**Comment by pedrobaeza** [comment]
I think the change can be included in the same PR as it's very related.

**Comment by bodedra** [comment]
@pedrobaeza @sergio-teruel It is ready for your review.

**Review by pedrobaeza — APPROVED** [review]
Thanks for discovering the problem

**Comment by bodedra** [comment]
@pedrobaeza @sergio-teruel Appreciate your support. This PR is :green_apple:

**Comment by bodedra** [comment]
We need to same thing for account.move.line object. https://github.com/OCA/margin-analysis/blob/92a40ed7afbcbfb85f18ef4a8dcfb5a86e5dec8a/account_invoice_margin/models/account_invoice.py#L87

**Comment by pedrobaeza** [comment]
Yes, in this case is to extract this to `account_invoice_margin_sale`:
```python
    @api.depends("purchase_price", "price_subtotal")
    def _compute_margin(self):
        for line in self:
            if (
                line.move_id
                and line.move_id.type[:2] == "in"
                or any(line.sale_line_ids.mapped("is_downpayment"))
            ):
                line.update(
                    {"margin": 0.0, "margin_signed": 0.0, "margin_percent": 0.0}
                )
            else:
                super(AccountMoveLine, line)._compute_margin()
```

**Review by sergio-teruel — APPROVED** [review]

**Comment by sergio-teruel** [comment]
Thanks @bodedra

**Review by marcelsavegnago — APPROVED** [review]

File: account_invoice_margin_sale/models/account_invoice.py [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+class AccountMoveLine(models.Model):
+    _inherit = "account.move.line"
+
+    @api.depends("purchase_price", "price_subtotal")
+    def _compute_margin(self):
+        for line in self:
+            if (
+                line.move_id
+                and line.move_id.type[:2] == "in"
+                or any(line.sale_line_ids.mapped("is_downpayment"))
```
Comment (pedrobaeza) [inline account_invoice_margin_sale/models/account_invoice.py]: I didn't see that the other condition is met in `super`, so no need to add it here:
```suggestion
                any(line.sale_line_ids.mapped("is_downpayment"))
```

Comment (bodedra) [reply account_invoice_margin_sale/models/account_invoice.py]: Addressed!

**Review by pedrobaeza — APPROVED** [review]
Thank you very much!

**Pull request merged**